### DISCUSSION TITLE
authenticated queries and subscriptions

### DISCUSSION THREAD

**Discussion by cezarneaga** [urql-graphql/urql#1065] [Q&A] [answered]
Hi there,

i have setup the following:
```
const addAuthToOperation = ({ operation }) => {
    const fetchOptions =
      typeof operation.context.fetchOptions === 'function'
        ? operation.context.fetchOptions()
        : operation.context.fetchOptions || {}
    return {
      ...operation,
      context: {
        ...operation.context,
        fetchOptions: {
          ...fetchOptions,
          headers: {
                ...fetchOptions.headers,
                Authorization: `Bearer token...`,
          }
        },
      },
    }
  }
```

and the client:

```
const client = createClient({
    url,                     // this is the graphql server
    fetch,                // this is the isomorfic-fetch
    requestPolicy: 'cache-and-network',
    exchanges: [
      dedupExchange,
      cacheExchange,
      authExchange({
        getAuth,                          // this is a function that returns `token` and `refreshToken`
        addAuthToOperation,     // see above
        didAuthError,                  // just returns a graph err
      }),
      fetchExchange,
      subscriptionExchange({
        forwardSubscription(operation) {
          return subscriptionClient.request(operation)
        },
      }),
    ],
  })
```

in this scenario, is the subscription client also receiving the authentication?
i could only get it working by adding to the `subscriptionClient` the connectionParams

```
const subscriptionClient = new SubscriptionClient(
    wsUri, // this is the wss:// url
    {
      reconnect: true,
      connectionParams: {
        headers: { 
          Authorization: `Bearer token ...` 
        },
      },
    },
    ws // this is `isomorphic-ws`
  )
```

is there a way to pass the authExchange to subscriptionClient too?
thanks,
C

**Comment by JoviDeCroock** [comment, marked as answer]
Hey,

Adding the token in the `subscriptionExchange` is normal yes, in case you always want the possibly refreshed token you should pass a function to the `connectionParams`:

```
connectionParams?: Object | Function | Promise<Object>
```

This function should get it from localStorage, ...

Are you saying you want to refresh tokens on active subscriptions? This is something where our exchanges actually can't go. Some information about that can be found [here](https://github.com/apollographql/subscriptions-transport-ws/issues/293)

Every newly activated subscription will pass through the authExchange and refresh the token.

**Reply by cezarneaga** [reply]
thanks @JoviDeCroock! 
just wanted to make sure what i did was correct. 
i am having this issue where on a hard refresh the subscription doesnt work but if i navigate through to the page it does. its now clear it isnt because i was doing something wrong there.

thanks,
C

**Reply by JoviDeCroock** [reply]
Glad I could help out, feel free to reach out in case you run into other blockers. Marking this as the answer.

**Reply by 97amarnathk** [reply]
@JoviDeCroock I recently faced a similar issue related to refreshing authTokens, and I found [this](https://github.com/enisdenjo/graphql-ws#auth-token) reference implementation, which seems to work quite well.

I pass accessToken in `connectionParams`, and keep on checking it in the server inside `onConnect`, `onNext`, `onSubscribe` operations. If token is found to be expired, the socket is closed, and client will reconnect and resubscribe to all subscriptions

A minor problem in this approach is `How to ensure that all messages happening in the middle of this reconnect is delivered`?

Any idea on how I can solve this ?

**Reply by 97amarnathk** [reply]
https://github.com/FormidableLabs/urql/discussions/1598

A discussion I have opened for the same
